File: phase-1-dsa-foundations/module-2-core-data-structures/hashmaps-sets/MapSetOperations.java
import java.util.*;

/**
 * Demonstrating HashMap and HashSet operations.
 */
public class MapSetOperations {

    public static void main(String[] args) {
        demonstrateHashMap();
        demonstrateHashSet();
        demonstrateFrequencyCount();
    }

    private static void demonstrateHashMap() {
        System.out.println("--- HashMap Operations ---");
        // Key: Name, Value: Age
        Map<String, Integer> map = new HashMap<>();

        // 1. Put - O(1)
        map.put("Alice", 25);
        map.put("Bob", 30);
        map.put("Alice", 26); // Updates value
        System.out.println("Map: " + map);

        // 2. Get - O(1)
        System.out.println("Alice's Age: " + map.get("Alice"));
        System.out.println("Unknown's Age: " + map.getOrDefault("Charlie", -1));

        // 3. Contains - O(1)
        System.out.println("Has Bob? " + map.containsKey("Bob"));

        // 4. Iteration
        System.out.print("Entries: ");
        for (Map.Entry<String, Integer> entry : map.entrySet()) {
            System.out.print(entry.getKey() + "=" + entry.getValue() + " ");
        }
        System.out.println();
    }

    private static void demonstrateHashSet() {
        System.out.println("\n--- HashSet Operations ---");
        Set<String> set = new HashSet<>();

        // 1. Add - O(1)
        set.add("Apple");
        set.add("Banana");
        boolean added = set.add("Apple"); // Duplicate
        System.out.println("Added Duplicate Apple? " + added); // false
        System.out.println("Set: " + set);

        // 2. Operations
        set.remove("Banana");
        System.out.println("Has Banana? " + set.contains("Banana"));
    }

    // Common Interview Pattern
    private static void demonstrateFrequencyCount() {
        System.out.println("\n--- Frequency Count Pattern ---");
        String[] arr = { "apple", "banana", "apple", "orange", "banana", "apple" };
        Map<String, Integer> counts = new HashMap<>();

        for (String s : arr) {
            counts.put(s, counts.getOrDefault(s, 0) + 1);
        }

        System.out.println("Frequencies: " + counts);
    }
}

File: go/phase-1-dsa-foundations/module-2-core-data-structures/hashmaps-sets/map_set_operations.go
package main

import "fmt"

func DemonstrateMap() {
	fmt.Println("--- Map Operations ---")
	// Key: Name, Value: Age
	m := make(map[string]int)

	// 1. Put
	m["Alice"] = 25
	m["Bob"] = 30
	m["Alice"] = 26 // Update
	fmt.Printf("Map: %v\n", m)

	// 2. Get
	fmt.Printf("Alice's Age: %d\n", m["Alice"])
	
	// Get with check
	val, exists := m["Charlie"]
	if !exists {
		fmt.Printf("Charlie not found (default int: %d)\n", val)
	}

	// 3. Delete
	delete(m, "Bob")
	
	// 4. Iteration (Order is random!)
	fmt.Print("Entries: ")
	for k, v := range m {
		fmt.Printf("%s=%d ", k, v)
	}
	fmt.Println()
}

func DemonstrateSet() {
	fmt.Println("\n--- Set (map[T]bool) Operations ---")
	set := make(map[string]bool)

	// 1. Add
	set["Apple"] = true
	set["Banana"] = true
	set["Apple"] = true // Duplicate (no-op effectively)
	
	fmt.Printf("Set: %v\n", set)

	// 2. Check Exists
	if set["Banana"] {
		fmt.Println("Has Banana")
	}

	// 3. Delete
	delete(set, "Banana")
	fmt.Printf("Has Banana? %v\n", set["Banana"])
}

func DemonstrateFrequencyCount() {
	fmt.Println("\n--- Frequency Count Pattern ---")
	arr := []string{"apple", "banana", "apple", "orange", "banana", "apple"}
	counts := make(map[string]int)

	for _, s := range arr {
		counts[s]++
	}

	fmt.Printf("Frequencies: %v\n", counts)
}

func main() {
	DemonstrateMap()
	DemonstrateSet()
	DemonstrateFrequencyCount()
}
